File: Charms Web Tools/Image Tools/Noise Studio/Old Noise Studio Versions/digital-grain-studio-modular v25 test/extract.mjs
import fs from 'fs';
import path from 'path';

const srcDir = './src';
const dirs = ['state', 'ui', 'webgl', 'shaders', 'effects', 'workers'];

// Create directories
fs.mkdirSync(srcDir, { recursive: true });
for (const d of dirs) {
    fs.mkdirSync(path.join(srcDir, d), { recursive: true });
}

const inputPath = '../v25 unbundled.html';
const source = fs.readFileSync(inputPath, 'utf8');

// 1. Extract CSS
const styleStart = source.indexOf('<style>');
const styleEnd = source.indexOf('</style>');
let css = source.substring(styleStart + 7, styleEnd);
fs.writeFileSync(path.join(srcDir, 'style.css'), css.trim() + '\n');

// 2. Extract HTML UI
const firstScript = source.indexOf('<script'); // Line 2412
let htmlHeader = source.substring(0, styleStart); // up to <style>
let htmlBody = source.substring(styleEnd + 8, firstScript); // after </style> up to <script

let finalHtml = htmlHeader + '    <link rel="stylesheet" href="/src/style.css">\n' + htmlBody;
finalHtml += '    <script type="module" src="/src/main.js"></script>\n</body>\n</html>\n';
fs.writeFileSync('index.html', finalHtml);

// 3. Extract Shaders
const shaderRegex = /<script\s+id="([^"]+)"\s+type="x-shader\/x-(fragment|vertex)">([\s\S]*?)<\/script>/g;
let match;
let count = 0;
while ((match = shaderRegex.exec(source)) !== null) {
    const id = match[1];
    const type = match[2];
    const code = match[3];
    const ext = type === 'fragment' ? '.frag' : '.vert';
    fs.writeFileSync(path.join(srcDir, 'shaders', id + ext), code.trim() + '\n');
    count++;
}

// 4. Extract Main Logic (Everything not a shader)
const jsRegex = /<script(?:[^>]*?)>([\s\S]*?)<\/script>/g;
let jsContent = '';
let jsMatch;
while ((jsMatch = jsRegex.exec(source)) !== null) {
    if (jsMatch[0].includes('x-shader')) continue;
    jsContent += jsMatch[1] + '\n';
}
fs.writeFileSync(path.join(srcDir, 'legacy_main.js'), jsContent);

console.log(`Extracted style.css, index.html, legacy_main.js, and ${count} shaders.`);

File: Charms Web Tools/Image Tools/Noise Studio/Old Noise Studio Versions/digital-grain-studio-modular v25 test/extract.py
import os
import re

src_dir = './src'
dirs = ['state', 'ui', 'webgl', 'shaders', 'effects', 'workers']

for d in dirs:
    os.makedirs(os.path.join(src_dir, d), exist_ok=True)

input_path = '../v25 unbundled.html'
with open(input_path, 'r', encoding='utf-8') as f:
    source = f.read()

# 1. Extract CSS
style_start = source.find('<style>')
style_end = source.find('</style>')
css = source[style_start + 7:style_end].strip()
with open(os.path.join(src_dir, 'style.css'), 'w', encoding='utf-8') as f:
    f.write(css + '\n')

# 2. Extract HTML UI
first_script = source.find('<script')
html_header = source[:style_start]
html_body = source[style_end + 8:first_script]

final_html = html_header + '    <link rel="stylesheet" href="/src/style.css">\n' + html_body
final_html += '    <script type="module" src="/src/main.js"></script>\n</body>\n</html>\n'
with open('index.html', 'w', encoding='utf-8') as f:
    f.write(final_html)

# 3 & 4. Parse all script tags manually
js_content = ''
shader_count = 0

script_starts = [m.start() for m in re.finditer(r'<script', source)]

for start in script_starts:
    tag_end = source.find('>', start)
    content_end = source.find('</script>', tag_end)
    
    if tag_end == -1 or content_end == -1:
        continue
        
    tag = source[start:tag_end+1]
    content = source[tag_end+1:content_end]
    
    if 'x-shader' in tag:
        # Extract ID
        id_match = re.search(r'id=["\']([^"\']+)["\']', tag)
        type_match = re.search(r'type=["\']([^"\']+)["\']', tag)
        
        script_id = id_match.group(1) if id_match else f"shader_{shader_count}"
        script_type = type_match.group(1) if type_match else ""
        
        ext = '.vert' if 'vertex' in script_type else '.frag'
        
        with open(os.path.join(src_dir, 'shaders', script_id + ext), 'w', encoding='utf-8') as f:
            f.write(content.strip() + '\n')
        shader_count += 1
    else:
        js_content += content + '\n'

with open(os.path.join(src_dir, 'legacy_main.js'), 'w', encoding='utf-8') as f:
    f.write(js_content)

print(f"Extracted style.css, index.html, legacy_main.js, and {shader_count} shaders.")
